tracks/kotlin/exercises/rotational-cipher/mentoring.md

```diff
@@ -0,0 +1,42 @@
+### Reasonable Solutions
+
+```kotlin
+class RotationalCipher(private val key: Int) {
+    init {
+        require(key in 0..26) {
+            "Key should be integer between 0 and 26"
+        }
+    }
+
+    fun encode(input: String): String =
+            input.map(shiftBy(key)).joinToString("")
+
+    // Do not allow private members to access object state.
+    // This allows for easier refactoring to another object
+    // because their state is self contained.
+    private fun shiftBy(key: Int): (Char) -> Char = {
+        if (shouldNotShift(it)) it
+        else doShift(key, it)
+    }
+
+    private fun doShift(key: Int, c: Char): Char {
+        val mod = 26
+        val shift = if (c.isUpperCase()) {
```

Comment:
Kotlin's official style guide suggests no brackets and inline for this:

```kotlin
val shift = if (c.isUpperCase()) 65 else 97
```

Reply:
I prefer these magic numbers (`mod`, `65` and `97`) to be private constants on the `companion object`.

Reply:
Would you still accept the pull without the private constants for now to get more input?

I prefer them here because the context within which they sit explain their use clearly enough without another abstraction. Also, they are only used here so adding them as consts on the companion object increase the scope in which they are visible, adding to the mental overhead of where I care about them.

Reply:
I replied by email, but yes. It's fine for now. The biggest issue is that there is no context. You and I know these are A and Z. But that's not apparent enough. 

However you are right that having something that's not the best but still good is better than having nothing at all!